primitives/src/lib.rs

```diff
@@ -104,3 +104,141 @@ mod prelude {
+    Decode(<T::Decoder as Decoder>::Error),
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+impl<T: Decodable> fmt::Debug for ParsePrimitiveError<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::OddLengthString(odd_err) => write_err!(f, "odd length string"; odd_err),
+            Self::InvalidChar(char_err) => write_err!(f, "invalid character"; char_err),
+            // Decoder error types don't have Debug, so we only provide this generic error
+            Self::Decode(_) => write!(f, "failure decoding hex string into {}", core::any::type_name::<T>()),
+        }
+    }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+impl<T: Decodable> fmt::Display for ParsePrimitiveError<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { fmt::Debug::fmt(&self, f) }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+impl<T: Decodable> From<hex::DecodeVariableLengthBytesError> for ParsePrimitiveError<T> {
+    fn from(dec_err: hex::DecodeVariableLengthBytesError) -> Self {
+        use hex::DecodeVariableLengthBytesError as D;
+
+        match dec_err {
+            D::InvalidChar(err) => Self::InvalidChar(err),
+            D::OddLengthString(err) => Self::OddLengthString(err),
+        }
+    }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+impl<T: Decodable> From<convert::Infallible> for ParsePrimitiveError<T> {
+    fn from(never: convert::Infallible) -> Self { match never {} }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc", feature = "std"))]
+impl<T: Decodable> std::error::Error for ParsePrimitiveError<T> {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::OddLengthString(ref e) => Some(e),
+            Self::InvalidChar(ref e) => Some(e),
+            Self::Decode(_) => None,
+        }
+    }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+pub(crate) mod hex_codec {
+    use super::{fmt, Decodable, ParsePrimitiveError};
+
+    use encoding::{Encodable, EncodableByteIter};
+    use hex_unstable::{BytesToHexIter, Case};
+
+    /// Writes an Encodable object to the given formatter in the requested case.
+    #[inline]
+    fn hex_write_tx_with_case<T: Encodable + Decodable>(obj: &HexPrimitive<T>, f: &mut fmt::Formatter, case: Case) -> fmt::Result {
+        let iter = BytesToHexIter::new(
+            encoding::EncodableByteIter::new(obj.0),
+            case
+        );
+        let collection = iter.collect::<alloc::string::String>();
+        f.pad(&collection)
+    }
+
+    /// Hex encoding wrapper type for Encodable + Decodable types.
+    ///
+    /// Provides default implementations for `Display`, `Debug`, `LowerHex`, and `UpperHex`.
+    /// Also provides [`Self::from_str`] for parsing a string to a `T`.
+    /// This can be used to implement hex display traits for any encodable types.
+    pub struct HexPrimitive<'a, T: Encodable + Decodable>(pub &'a T);
```

Comment:
```suggestion
    pub(crate) struct HexPrimitive<'a, T: Encodable + Decodable>(pub &'a T);
```

Not that it really matters but just for clarity.